### ISSUE TITLE
Leader: GC can cause a batch to remain in Closing state indefinitely

### ISSUE THREAD

**Issue by branlwyd** [divviup/janus#2443]
When writing an aggregation job back with `aggregation_job_writer.rs`, we close relevant batches if all of the following conditions hold:

(a) the batch is already in `Closing` state
(b) there are no outstanding aggregation jobs
(c) there are no remaining unaggregated reports in the interval

These conditions are evaluated only during aggregation-job writes. But in the presence of GC, (c) can change over time if reports are GC'ed. This could lead to batches being forever stuck in `Closing` state.

**Comment by branlwyd** [comment]
Idea for time-interval: tweak client-report GC such that client reports for a given `time_precision` window all become eligible for GC once the end of the window passes beyond the GC horizon. If we made that change, batches would be GC'ed at the same time as all reports in the batch.

One possible downside of this idea is that reports would now become eligible for GC in large groups, rather than continuously; this might cause poor GC behavior, especially for tasks with a large `time_precision`.

I think fixed-size doesn't have the problem described in this issue, since reports are attached to batches at the point of being added to an aggregation job. Therefore, condition (c) is irrelevant for fixed-size.

**Comment by branlwyd** [comment]
> One possible downside of this idea is that reports would now become eligible for GC in large groups, rather than continuously; this might cause poor GC behavior, especially for tasks with a large time_precision.

Since clients will typically be rounding their `client_timestamp` values to a multiple of `time_precision` already, this isn't actually a downside. (clients which don't round their timestamps will still be affected, but will likely be rare.

**Comment by branlwyd** [comment]
One thing to be careful of with the idea suggested in https://github.com/divviup/janus/issues/2443#issuecomment-1879528048: I'm not sure the math required to perform this check would be amenable to finding GC'able reports via a `client_timestamp` index. But I'm also not sure Postgres isn't smart enough to somehow still use an index. We should EXPLAIN what the client-report GC query looks like to gain insight here.

**Comment by branlwyd** [comment]
The concern in https://github.com/divviup/janus/issues/2443#issuecomment-1889943182 proved to be real: I hacked together a partial implementation of the GC-everything-in-a-`time_precision`-window-together idea, and the client-GC selection query switched from an index scan to a sequential scan.

Old query plan:
```
 Limit  (cost=0.58..84219.75 rows=5000 width=8)
   ->  Nested Loop  (cost=0.58..4063609.18 rows=241252 width=8)
         ->  Seq Scan on tasks  (cost=0.00..12.38 rows=1 width=16)
               Filter: (task_id = '\xc8be6ada53cb4e5d55807bcc114041f0112e9e67666fbfbb40a891c48d287135'::bytea)
         ->  Index Scan using client_reports_task_and_timestamp_index on client_reports  (cost=0.58..4025398.58 rows=3819822 width=24)
               Index Cond: ((task_id = tasks.id) AND (client_timestamp < COALESCE((now() - ((tasks.report_expiry_age)::double precision * '00:00:01'::interval)), ('-infinity'::timestamp without time zone)::timestamp with time zone)))
```

New query plan:
```
 Limit  (cost=1002.38..146973.01 rows=5000 width=8)
   ->  Gather  (cost=1002.38..7044143.90 rows=241252 width=8)
         Workers Planned: 2
         ->  Hash Join  (cost=2.38..7019018.70 rows=100522 width=8)
               Hash Cond: (client_reports.task_id = tasks.id)
               Join Filter: (to_timestamp((((floor(('1'::numeric + (EXTRACT(epoch FROM client_reports.client_timestamp) / (tasks.time_precision)::numeric))) * (tasks.time_precision)::numeric) + (tasks.time_precision)::numeric))::double precision) < COALESCE((now() - ((tasks.report_expiry_age)::double precision * '00:00:01'::interval)), ('-infinity'::timestamp without time zone)::timestamp with time zone))
               ->  Parallel Seq Scan on client_reports  (cost=0.00..6863220.33 rows=57297333 width=24)
               ->  Hash  (cost=2.36..2.36 rows=1 width=24)
                     ->  Index Scan using task_id_index on tasks  (cost=0.14..2.36 rows=1 width=24)
                           Index Cond: (task_id = '\xc8be6ada53cb4e5d55807bcc114041f0112e9e67666fbfbb40a891c48d287135'::bytea)
```

**Comment by branlwyd** [comment]
Collection job polling (implemented in #2782) fixes this: a GC'ed client report will be noticed with the next poll of the collection job for readiness.
